### PR TITLE
[bugfix] Resolve lost reference to meta object in `ResultMetricCollection`

### LINKED ISSUE #8918: self.log validation_step Misconfiguration error
## 🐛 Bug

After an update from 1.3.0 to 1.4.2, existing code breaks with:

```
  File "/home/rongcuid/miniconda3/envs/pytorch/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/logger_connector/result.py", line 464, in log
    raise MisconfigurationException(
pytorch_lightning.utilities.exceptions.MisconfigurationException: You called `self.log(loss/validate, ...)` twice in `validation_step` with different arguments. This is not allowed
```
There is only one `self.log` statement in my validation loop. However, there is an identical counterpart in the training step, which uses "loss/train" as name. It worked in 1.3.

### Expected behavior

Logs validation loss correctly.

### Environment

```
* CUDA:
	- GPU:
		- NVIDIA GeForce GTX 1080
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.21.1
	- pyTorch_debug:     False
	- pyTorch_version:   1.9.0
	- pytorch-lightning: 1.4.2
	- tqdm:              4.62.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.8.11
	- version:           #21~1626191760~20.04~55de9c3-Ubuntu SMP Wed Jul 21 20:31:55 UTC
```

### Additional context

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR resolves a bug where the meta object was lost during ResultCollection.to function. 

I removed entirely the usage of Metadata through the __init__ function as this was definitely a wrong pattern.

Now, the meta object is resolved directly from the children.

Fixes #8918

### Does your PR introduce any breaking changes? If yes, please list them.

No.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 1]
```diff
@@ -289,9 +289,12 @@ class ResultMetricCollection(dict):
...
     """
 
-    def __init__(self, *args, metadata: Optional[_Metadata] = None) -> None:
+    def __init__(self, *args) -> None:
         super().__init__(*args)
-        self.meta = metadata
+
+    @property
+    def meta(self) -> _Metadata:
+        return list(self.values())[0].meta
```
Comment (justusschock) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:297]: do we need a sanity check that meta from all values is the same?

**Review by justusschock — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:297]: No, the meta is automaticly set by us

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 2]
```diff
@@ -590,3 +590,29 @@ def get_metrics_at_idx(idx):
...
+def test_logging_dict_on_validation_step(tmpdir):
+    class TestModel(BoringModel):
+        def validation_step(self, batch, batch_idx):
+            loss = super().validation_step(batch, batch_idx)
+            loss = loss["x"]
+            metrics = {
+                "loss": loss,
+                "loss_1": loss,
+            }
+            self.log("val_metrics", metrics)
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:604]: You can repro with dummy values


```suggestion
            self.log("val_metrics", {"a": 1, "b": 2})
```

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 3]
```diff
@@ -590,3 +590,29 @@ def get_metrics_at_idx(idx):
...
+
+        validation_epoch_end = None
+
+    model = TestModel()
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=2,
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:612]: 1 is enough

```suggestion
        limit_train_batches=1,
```

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 4]
```diff
@@ -590,3 +590,29 @@ def get_metrics_at_idx(idx):
...
+        limit_val_batches=2,
+        max_epochs=2,
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:614]:
```suggestion
        num_sanity_val_batches=2,
        max_epochs=1,
```
To be explicit about what causes the bug

**Review by carmocca — APPROVED** [review]

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 2 above)
Comment (tchaton) [reply tests/trainer/logging_/test_eval_loop_logging.py:604]: Done !

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 3 above)
Comment (tchaton) [reply tests/trainer/logging_/test_eval_loop_logging.py:612]: Done !

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 4 above)
Comment (tchaton) [reply tests/trainer/logging_/test_eval_loop_logging.py:614]: This line https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/logger_connector/result.py#L595 + default metadata=None in ResultMetricCollection.

As apply_collection was re-creating the object and metadata wasn't passed to the init function, it just override itself with metadata and broke on next self.log.

I believe it is safer to removed entirely and depend on its children.
